src/charm.py

```diff
@@ -0,0 +1,192 @@
+        )
+        self.kube_control.create_kubeconfig(
+            self.CA_CERT_PATH, "/home/ubuntu/.kube/config", "ubuntu", self.unit.name
+        )
+        return True
+
+    def _check_certificates(self, event: EventBase) -> bool:
+        self.unit.status = MaintenanceStatus("Evaluating certificates.")
+        evaluation = self.certificates.evaluate_relation(event)
+        if evaluation:
+            if "Waiting" in evaluation:
+                self.unit.status = WaitingStatus(evaluation)
+            else:
+                self.unit.status = BlockedStatus(evaluation)
+            return False
+        self.CA_CERT_PATH.write_text(self.certificates.ca)
+        return True
+
+    def _check_config(self, event: EventBase) -> None:  # TODO: Not the best name
+        if not self._check_certificates(event):
+            return
+
+        if not self._check_kube_control(event):
+            return
+
+        channel = self.config.get("channel")
+        self._install_snaps(channel)
+
+        self.unit.status = ops.ActiveStatus("Ready to test.")
+
+    def _upgrade_charm(self, event: EventBase) -> None:
+        channel = self.config.get("channel")
+        self._install_snaps(channel)
+
+    def _install_snaps(self, channel: Optional[str]) -> None:
+        self.unit.status = ops.MaintenanceStatus("Installing core snap.")
+        self._ensure_snap("core")
+
+        self.unit.status = ops.MaintenanceStatus("Installing kubectl snap.")
+        self._ensure_snap("kubectl", channel=channel)
+
+        self.unit.status = ops.MaintenanceStatus("Installing kubernetes-test snap.")
+        self._ensure_snap("kubernetes-test", channel=channel, classic=True)
+
+    def _ensure_snap(
+        self,
+        name: str,
+        state: SnapState = SnapState.Latest,
+        channel: Optional[str] = "",
+        classic: Optional[bool] = False,
+    ) -> None:
+        if not isinstance(name, str) or name == "":
+            raise ValueError("A name is required to ensure a snap.")
+
+        snap = self.snap_cache[name]
+        if not snap.present:
+            snap.ensure(state=state, classic=classic, channel=channel)
+
+    def _on_start(self, event: EventBase) -> None:
+        self.unit.status = ops.ActiveStatus()
+
+    def _check_kube_config(self, event: ActionEvent) -> bool:
+        if not pathlib.Path(KUBE_CONFIG_PATH).exists():
+            event.fail("Missing Kubernetes configuration. See logs for info.")
+            logger.error("Relate to the certificate authority and kubernetes-control-plane.")
+            return False
+        return True
+
+    def _log_has_errors(self, event: ActionEvent) -> bool:
+        action_uuid = os.getenv("JUJU_ACTION_UUID")
+
+        log_file_path = pathlib.Path(f"{action_uuid}.log")
+
+        if not log_file_path.exists():
+            msg = f"Logfile not found at expected location {log_file_path}"
+            logger.error(msg)
+            event.fail(msg)
+            return False
```

Comment:
Wanted to mention [`event.log(...)`](https://ops.readthedocs.io/en/latest/#ops.ActionEvent.log) here in case you need this.

Also -- i'm not sure this is necessary is it?  Are you just making sure the e2e snap creates this Path?

if so i think it would be in `/home/ubuntu/{action_uuid}.log` right? not on the juju agent's `cwd` which is down in `/var/lib/juju/agents/unit-kubernetes-e2e-N/charm`